src/analysis/processing/qgsalgorithmintersection.cpp

```diff
@@ -0,0 +1,268 @@
+/***************************************************************************
+  qgsalgorithmintersection.cpp
+  ---------------------
+  Date                 : April 2018
+  Copyright            : (C) 2018 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsalgorithmintersection.h"
+
+#include "qgsgeometrycollection.h"
+#include "qgsgeometryengine.h"
+
+#include "qgsmultipoint.h"
+#include "qgsmultipolygon.h"
+#include "qgsmultilinestring.h"
+
+///@cond PRIVATE
+
+static QgsGeometry geometryCollectionToSingleGeometryType( const QgsGeometry &geom, QgsWkbTypes::GeometryType geomType )
+{
```

Comment:
How do you feel about moving this to QgsGeometry? I think it's useful outside this alg. Also if it were in QgsGeometry you could likely avoid the clones by doing similar tricks to QgsGeometry::convertToMultipart.